### ISSUE TITLE
[java] DU false positive in DataflowAnomalyAnalysis

### ISSUE THREAD

**Issue by mcandre** [pmd/pmd#1606]
**Affects PMD Version:** 

3.9+

**Rule:** [DataflowAnomalyAnalysis](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#dataflowanomalyanalysis)

**Description:**

PMD's anomaly analysis gets easily confused, such as reporting the variable A as DU (undefined). The message is not structured to effectively communicate to the reader what the problem is or what should be done to resolve it. Furthermore, the message is not structured in a way to make the warning type easy to Google.

Here is an example bit of source that can trigger the rule:

```java
public final class Fib {
    /** utility class */
    private Fib() {}

    public static int fib(final int n) {
        Preconditions.checkArgument(n >= 0);

        if (n < 2) {
            return n;
        } else {
            int a = 0;
            int b = 1;
            final int m = n - 1;

            for (int i = 0; i < m; i++) {
                final int c = a;
                a = b;
                b = c + b;
            }

            return b;
        }
    }
}
```

Result:

```
n/a:11:	Found 'DU'-anomaly for variable 'a' (lines '11'-'23').
n/a:17:	Found 'DU'-anomaly for variable 'a' (lines '17'-'23').
```

I think the fact that A and B are reassigned during the iterative Fibonacci implementation is what is confusing the anomaly anlyzer.

As a workaround, I could exclude the DataflowAnomalyAnalysis rule from the errorprone.xml rule element, but instead I am opting to omit the entire `errorprone.xml` collection from my rulesets.xml:

```xml
<?xml version="1.0"?>
<ruleset name="Functioning PMD rules for Java"
    xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 https://pmd.sourceforge.io/ruleset_2_0_0.xsd">
    <description>Functioning PMD rules for Java</description>

    <rule ref="category/java/bestpractices.xml" />
    <rule ref="category/java/codestyle.xml" />
    <rule ref="category/java/design.xml" />
    <rule ref="category/java/documentation.xml" />
    <rule ref="category/java/multithreading.xml" />
    <rule ref="category/java/performance.xml" />
    <rule ref="category/java/security.xml" />
</ruleset>
```

**Comment by adangel** [comment]
The rule [DataflowAnomalyAnalysis](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#dataflowanomalyanalysis) has been deprecated in favour of [UnusedAssignment](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#unusedassignment), which is available with PMD 6.26.0.

Therefore this issue wont be fixed.

See also #2647.
